### PR TITLE
Fix GCStress 0xC on Unix

### PR DESCRIPTION
**Pull request by janvorli**
This change fixes GCStress 0xC on Unix that was not working due to the way that
we check that a specific IP is in managed code. To make that check safe,
a precondition checking that the thread is in cooperative GC mode was added
some time ago. But that causes a problem with the GCStress 0xC and PInvoke
stubs. The PInvoke stub switches to preemptive mode at some point and when
the code beyond that point is instrumented and the corresponding hardware
exception occurs, the exception address is not evaluated as being in managed
code and so the exception terminates the application.
This fix excludes the safety check of GC mode when we are running with GC stress
mode 8 or 4 (or both -> 0xc).

### CODE REVIEW

**Comment by janvorli** [comment]
@jkotas can you take a look please?

**Comment by janvorli** [comment]
cc: @gkhanna79 - the fact that we had the wrong casing of complus_gcstress in the test script that was fixed yesterday was hiding the fact that the gc stress mode was not working at all. It was crashing any app during its startup.

**Comment by gkhanna79** [comment]
@janvorli Thank you for fixing this. It is a bit unfortunate this was identified at this point in time. Does this also mean that the issue is unique to non-Windows due to our custom exception dispatch implementation? I am trying to understand why we didnt see this issue on Windows?

**Comment by gkhanna79** [comment]
Based upon your description, I believe we will need to take this change to release/1.0.0 branch as well.

**Comment by janvorli** [comment]
@gkhanna79 Yes, I think so.

File: src/vm/exceptionhandling.cpp [hunk 1]
```diff
@@ -5048,19 +5048,32 @@ bool IsDivByZeroAnIntegerOverflow(PCONTEXT pContext)
...
 
-BOOL PALAPI IsSafeToHandleHardwareException(PCONTEXT contextRecord, PEXCEPTION_RECORD exceptionRecord)
+BOOL IsManagedCode(PCODE controlPc)
 {
     Thread *pThread = GetThread();
+
+    // The ExecutionManager::IsManagedCode can be safely called in cooperative GC mode only.
+    // Unfortunately, when running GC stress mode that invokes GC after every jitted or NGENed
+    // instruction, we need to relax that to enable instrumentation of PInvoke stubs that switch to
+    // preemtive GC mode at some point.
```
Comment (jkotas) [inline src/vm/exceptionhandling.cpp]: Type: `preemtive` -> `preemptive`

File: src/vm/exceptionhandling.cpp [hunk 2]
```diff
@@ -5048,19 +5048,32 @@ bool IsDivByZeroAnIntegerOverflow(PCONTEXT pContext)
 }
 #endif //_AMD64_
 
-BOOL PALAPI IsSafeToHandleHardwareException(PCONTEXT contextRecord, PEXCEPTION_RECORD exceptionRecord)
+BOOL IsManagedCode(PCODE controlPc)
```
Comment (jkotas) [inline src/vm/exceptionhandling.cpp]: I think the original name was better matching what this function was doing.

Comment (jkotas) [inline src/vm/exceptionhandling.cpp]: (Also, there is `ExecutionManager::IsManagedCode` already...)

Comment (janvorli) [inline src/vm/exceptionhandling.cpp]: This is not a rename, there still is the IsSafeToHandleHardwareException

Comment (jkotas) [inline src/vm/exceptionhandling.cpp]: I see ... we should give it a better name as we have discussed over Skype.

**Comment by jkotas** [comment]
LGTM modulo comments.

**Comment by janvorli** [comment]
@gkhanna79 - this is due to the way we handle hardware exceptions on Unix and is not related to Windows.

**Comment by janvorli** [comment]
@jkotas I've modified it according to your feedback.

**Comment by adityamandaleeka** [comment]
@jkotas LGTM, but does this mean we reintroduce the possibility of deadlock when calling the ExecutionManager (in cases where GCStress is enabled)?

**Comment by janvorli** [comment]
@adityamandaleeka - yes, that's true when GCStress modes 8 and/or 4 are enabled. But after discussing it with @jkotas, I unfortunately don't see a reasonable way to handle that case better.

**Comment by jkotas** [comment]
GCStress modes 8 and 4 are not designed to be 100% reliable. They are designed to find hard-to-detect bugs, at the cost of having some noise. There are number of shortcuts similar to this one.

**Comment by jkotas** [comment]
cc @swaroop-sridhar since this is related to the `m_fPreemptiveGCDisabledForGCStress` issues he has been running into.

**Comment by sergiy-k** [comment]
LGTM

**Pull request merged**